File: solutions/zig/kindergarten-garden/1/kindergarten_garden.zig
const std = @import("std");

pub const Plant = enum { clover, grass, radishes, violets };

inline fn plant(ch: u8) Plant {
    return switch (ch) {
        'G' => .grass,
        'C' => .clover,
        'R' => .radishes,
        'V' => .violets,
        else => unreachable,
    };
}

pub fn plants(diagram: []const u8, student: []const u8) [4]Plant {
    var result: [4]Plant = undefined;
    const index = 2 * (student[0] - 'A');

    var off: usize = 0;
    var it = std.mem.splitScalar(u8, diagram, '\n');
    while (it.next()) |line| {
        result[off] = plant(line[index]);
        result[off + 1] = plant(line[index + 1]);
        off += 2;
    }
    return result;
}

File: solutions/c/kindergarten-garden/1/kindergarten_garden.c
#include "kindergarten_garden.h"

#include <assert.h>

static plant_t plant(char c) {
   switch (c) {
      case 'C': return CLOVER;
      case 'G': return GRASS;
      case 'R': return RADISHES;
      case 'V': return VIOLETS;
   }
   return -1;
}

plants_t plants(const char* diagram, const char* student) {
   char* p = (char*)diagram;
   while (p && *p != '\n') p++;
   assert(p); p++; assert(p);

   const int index = 2 * (student[0] - 'A');
   const plants_t result = {
      .plants = {
         plant(diagram[index]),
         plant(diagram[index+1]),
         plant(*(p+index)),
         plant(*(p+index+1)),
      }
   };
   return result;
}
